man/man1/zpios.1
'\" t
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\"
.\" Copyright 2013 Darik Horn <dajhorn@vanadac.com>. All rights reserved.
.\"
.\" Copyright (c) 2015, Intel Corporation.
.\"
.TH zpios 1 "2013 FEB 28" "ZFS on Linux" "User Commands"

.SH NAME
zpios \- Directly test the DMU.
.SH SYNOPSIS
.LP
.BI "zpios [options] <\-p " pool ">"

.SH DESCRIPTION
This utility runs in-kernel DMU performance and stress tests that do
not depend on the ZFS Posix Layer ("ZPL").

.SH OPTIONS
.HP
.BI "\-t" " regex" ", \-\-threadcount" " regex"
.IP
Start this many threads for each test series, specified as a comma
delimited regular expression. (eg: "-t 1,2,3")
.IP
This option is mutually exclusive with the \fBthreadcount_*\fR
options.
.HP
.BI "\-l" " regex_low" ", \-\-threadcount_low" " regex_low"
.HP
.BI "\-h" " regex_high" ", \-\-threadcount_high" " regex_high"
.HP
.BI "\-e" " regex_incr" ", \-\-threadcount_incr" " regex_incr"
.IP
Start \fIregex_low\fR threads for the first test,
add \fIregex_incr\fR threads for each subsequent test,
and start \fIregex_high\fR threads for the last test.
.IP
These three options must be specified together and are mutually
exclusive with the \fBthreadcount\fR option.
.HP
.BI "\-n" " regex" ", \-\-regioncount" " regex"
.IP
Create this many regions for each test series, specified as a comma
delimited regular expression. (eg: "-n 512,4096,65536")
.IP
This option is mutually exclusive with the \fBregioncount_*\fR
options.
.HP
.BI "\-i" " regex_low" ", \-\-regioncount_low" " regex_low"
.HP
.BI "\-j" " regex_high" ", \-\-regioncount_high" " regex_high"
.HP
.BI "\-k" " regex_incr" ", \-\-regioncount_incr" " regex_incr"
.IP
Create \fIregex_low\fR regions for the first test,
add \fIregex_incr\fR regions for each subsequent test, and
create \fIregex_high\fR regions for the last test.
.IP
These three options must be specified together and are mutually
exclusive with the \fBregioncount\fR option.
.HP
.BI "\-o" " size" ", \-\-offset" " size"
.IP
Create regions at \fIsize\fR offset for each test series, specified as
a comma delimited regular expression with an optional unit suffix.
(eg: "-o 4M" means four megabytes.)
.IP
This option is mutually exclusive with the \fBoffset_*\fR options.
.HP
.BI "\-m" " size_low" ", \-\-offset_low" " size_low"
.HP
.BI "\-q" " size_high" ", \-\-offset_high" " size_high"
.HP
.BI "\-r" " size_incr" ", \-\-offset_incr" " size_incr"
.IP
Create a region at \fIsize_low\fR offset for the first test, add
\fIsize_incr\fR to the offset for each subsequent test, and create
a region at \fIsize_high\fR offset for the last test.
.IP
These three options must be specified together and are mutually
exclusive with the \fBoffset\fR option.
.HP
.BI "\-c" " size" ", \-\-chunksize" " size"
.IP
Use \fIsize\fR chunks for each test, specified as a comma delimited
regular expression with an optional unit suffix. (eg: "-c 1M" means
one megabyte.) The chunk size must be at least the region size.
.IP
This option is mutually exclusive with the \fBchunksize_*\fB options.
.HP
.BI "\-a" " size_low" ", \-\-chunksize_low" " size_low"
.HP
.BI "\-b" " size_high" ", \-\-chunksize_high" " size_high"
.HP
.BI "\-g" " size_incr" ", \-\-chunksize_incr" " size_incr"
.IP
Use a \fIsize_low\fR chunk size for the first test, add \fIsize_incr\fR
to the chunk size for each subsequent test, and use a \fIsize_high\fR
chunk size for the last test.
.IP
These three options must be specified together and are mutually
exclusive with the \fBchunksize\fR option.
.HP
.BI "\-s" " size" ", \-\-regionsize" " size"
.IP
Use \fIsize\fR regions for each test, specified as a comma delimited
regular expression with an optional unit suffix. (eg: "-s 1M" means
one megabyte.)
.IP
This option is mutually exclusive with the \fBregionsize_*\fB options.
.HP
.BI "\-A" " size_low" ", \-\-regionsize_low" " size_low"
.HP
.BI "\-B" " size_high" ", \-\-regionsize_high" " size_high"
.HP
.BI "\-C" " size_incr" ", \-\-regionsize_incr" " size_incr"
.IP
Use a \fIsize_low\fR region size for the first test, add \fIsize_incr\fR
to the region size for each subsequent test, and use a \fIsize_high\fR
region size for the last test.
.IP
These three options must be specified together and are mutually
exclusive with the \fBregionsize\fR option.
.HP
.BI "\-S" " size | sizes" ", \-\-blocksize" " size | sizes"
.IP
Use \fIsize\fR ZFS blocks for each test, specified as a comma delimited
regular expression with an optional unit suffix. (eg: "-S 1M" means
one megabyte.) The supported range is powers of two from 128K through 16M.
A range of blocks can be tested as follows: "-S 128K,256K,512K,1M".
.IP
.HP
.BI "\-L" " dmu_flags" ", \-\-load" " dmu_flags"
.IP
Specify \fIdmuio\fR for regular DMU_IO, \fIssf\fR for single shared
file access, or \fIfpp\fR for per thread access. Use commas to delimit
multiple flags. (eg: "-L dmuio,ssf")
.HP
.BI "\-p" " name" ", \-\-pool" " name"
.IP
The pool name, which is mandatory.
.HP
.BI "\-M" " test" ", \-\-name" " test"
.IP
An arbitrary string that appears in the program output.
.HP
.BI "-x, \-\-cleanup"
.IP
Enable the DMU_REMOVE flag.
.HP
.BI "\-P" " command" ", \-\-prerun" " command"
.IP
Invoke \fIcommand\fR from the kernel before running the test. Shell
expansion is not performed and the environment is set to
HOME=/; TERM=linux; PATH=/sbin:/usr/sbin:/bin:/usr/bin.
.HP
.BI "\-R" " command" ", \-\-postrun" " command"
.IP
Invoke \fIcommand\fR from the kernel after running the test. Shell
expansion is not performed and the environment is set to
HOME=/; TERM=linux; PATH=/sbin:/usr/sbin:/bin:/usr/bin.
.HP
.BI "\-G" " directory" ", \-\-log" " directory"
.IP
Put logging output in this directory.
.HP
.BI "\-I" " size" ", \-\-regionnoise" " size"
.IP
Randomly vary the \fBregionsize\fR parameter for each test
modulo \fIsize\fR bytes.
.HP
.BI "\-N" " size" ", \-\-chunknoise" " size"
.IP
Randomly vary the \fBchunksize\fR parameter for each test
modulo \fIsize\fR bytes.
.HP
.BI "\-T" " time" ", \-\-threaddelay" " time"
.IP
Randomly vary the execution time for each test
modulo \fItime\fR kernel jiffies.
.HP
.BI "\-V" "" ", \-\-verify" ""
.IP
Enable the DMU_VERIFY flag for trivial data verification.
.HP
.BI "\-z" "" ", \-\-zerocopy" ""
.IP
Enable the DMU_READ_ZC and DMU_WRITE_ZC flags, which are
currently unimplemented for Linux.
.IP
.HP
.BI "\-O" "" ", \-\-nowait" ""
.IP
Enable the DMU_WRITE_NOWAIT flag.
.HP
.BI "\-f" "" ", \-\-noprefetch" ""
.IP
Enable the DMU_READ_NOPF flag.
.HP
.BI "\-H" "" ", \-\-human\-readable" ""
.IP
Print PASS and FAIL results explicitly and put unit suffixes on large
numbers.
.HP
.BI "\-v" "" ", \-\-verbose" ""
.IP
Increase output verbosity.
.HP
.BI "\-?" " " ", \-\-help" " "
.IP
Print the usage message.
.SH "AUTHORS"
The original zpios implementation was created by Cluster File Systems
Inc and adapted to ZFS on Linux by Brian Behlendorf
<behlendorf1@llnl.gov>.

This man page was written by Darik Horn <dajhorn@vanadac.com>.
.SH "SEE ALSO"
.BR zpool (8),
.BR zfs (8)


scripts/zpios.sh
#!/bin/bash
#
# Wrapper script for easily running zpios based tests
#

basedir="$(dirname "$0")"

SCRIPT_COMMON=common.sh
if [ -f "${basedir}/${SCRIPT_COMMON}" ]; then
. "${basedir}/${SCRIPT_COMMON}"
else
echo "Missing helper script ${SCRIPT_COMMON}" && exit 1
fi

PROG=zpios.sh
DATE=$(date +%Y%m%d-%H%M%S)
if [ "${ZPIOS_MODULES}" ]; then
	MODULES=(${ZPIOS_MODULES[*]})
else
	MODULES=(zpios)
fi

usage() {
cat << EOF
USAGE:
$0 [hvp] <-c config> <-t test>

DESCRIPTION:
        Helper script for easy zpios benchmarking.

OPTIONS:
        -h      Show this message
        -v      Verbose
        -f      Force everything
        -p      Enable profiling
        -c      Zpool configuration
        -t      Zpios test
        -o      Additional zpios options
        -l      Additional zpool options
        -s      Additional zfs options

EOF
}

unload_die() {
	unload_modules
	while [ -c /dev/zpios ]; do
		sleep 1
	done

	exit 1
}

print_header() {
	echo --------------------- ZPIOS RESULTS ----------------------------
	echo -n "Date: "; date
	echo -n "Kernel: "; uname -r
	dmesg | grep "Loaded Solaris Porting Layer" | tail -n1
	dmesg | grep "Loaded ZFS Filesystem" | tail -n1
	echo
}

print_spl_info() {
	echo --------------------- SPL Tunings ------------------------------
	"${SYSCTL}" -A | grep spl

	if [ -d /sys/module/spl/parameters ]; then
		grep "[0-9]" /sys/module/spl/parameters/*
	else
		grep "[0-9]" /sys/module/spl/*
	fi

	echo
}

print_zfs_info() {
	echo --------------------- ZFS Tunings ------------------------------
	"${SYSCTL}" -A | grep zfs

	if [ -d /sys/module/zfs/parameters ]; then
		grep "[0-9]" /sys/module/zfs/parameters/*
	else
		grep "[0-9]" /sys/module/zfs/*
	fi

	echo
}

print_stats() {
	echo ---------------------- Statistics -------------------------------
	${SYSCTL} -A | grep spl | grep stack_max

	if [ -d /proc/spl/kstat/ ]; then
		if [ -f /proc/spl/kstat/zfs/arcstats ]; then
			echo "* ARC"
			cat /proc/spl/kstat/zfs/arcstats
			echo
		fi

		if [ -f /proc/spl/kstat/zfs/vdev_cache_stats ]; then
			echo "* VDEV Cache"
			cat /proc/spl/kstat/zfs/vdev_cache_stats
			echo
		fi
	fi

	if [ -f /proc/spl/kmem/slab ]; then
		echo "* SPL SLAB"
		cat /proc/spl/kmem/slab
		echo
	fi

	echo
}

check_test() {

	if [ ! -f "${ZPIOS_TEST}" ]; then
		local NAME=$(basename "${ZPIOS_TEST}" .sh)
		ERROR="Unknown test '${NAME}', available tests are:\n"

		while read TST; do
			local NAME=$(basename "${TST}" .sh)
			ERROR="${ERROR}${NAME}\n"
		done < <(find "${ZPIOSDIR}" -maxdepth 1 -type f -name "*.sh")

		return 1
	fi

	return 0
}

zpios_profile_config() {
cat > "${PROFILE_DIR}/zpios-config.sh" << EOF
#
# Zpios Profiling Configuration
#

PROFILE_DIR=/tmp/zpios/${ZPOOL_CONFIG}+${ZPIOS_TEST_ARG}+${DATE}
PROFILE_PRE=${ZPIOSPROFILEDIR}/zpios-profile-pre.sh
PROFILE_POST=${ZPIOSPROFILEDIR}/zpios-profile-post.sh
PROFILE_USER=${ZPIOSPROFILEDIR}/zpios-profile.sh
PROFILE_PIDS=${ZPIOSPROFILEDIR}/zpios-profile-pids.sh
PROFILE_DISK=${ZPIOSPROFILEDIR}/zpios-profile-disk.sh
PROFILE_ARC_PROC=/proc/spl/kstat/zfs/arcstats
PROFILE_VDEV_CACHE_PROC=/proc/spl/kstat/zfs/vdev_cache_stats

OPROFILE_KERNEL="/boot/vmlinux-$(uname -r)"
OPROFILE_KERNEL_DIR="/lib/modules/$(uname -r)/kernel/"
OPROFILE_SPL_DIR=${SPLBUILD}/module/
OPROFILE_ZFS_DIR=${MODDIR}

EOF
}

zpios_profile_start() {
	PROFILE_DIR=/tmp/zpios/${ZPOOL_CONFIG}+${ZPIOS_TEST_ARG}+${DATE}

	mkdir -p "${PROFILE_DIR}"
	zpios_profile_config
	. "${PROFILE_DIR}/zpios-config.sh"

	ZPIOS_OPTIONS="${ZPIOS_OPTIONS} --log=${PROFILE_DIR}"
	ZPIOS_OPTIONS="${ZPIOS_OPTIONS} --prerun=${PROFILE_PRE}"
	ZPIOS_OPTIONS="${ZPIOS_OPTIONS} --postrun=${PROFILE_POST}"

	/usr/bin/opcontrol --init
	/usr/bin/opcontrol --setup --vmlinux="${OPROFILE_KERNEL}"
}

zpios_profile_stop() {
	/usr/bin/opcontrol --shutdown
	/usr/bin/opcontrol --deinit
}

PROFILE=
ZPOOL_CONFIG=zpool-config.sh
ZPIOS_TEST=zpios-test.sh
ZPOOL_NAME=zpios
ZPIOS_OPTIONS=
ZPOOL_OPTIONS=""
ZFS_OPTIONS=""

while getopts 'hvfpc:t:o:l:s:' OPTION; do
	case $OPTION in
	h)
		usage
		exit 1
		;;
	v)
		VERBOSE=1
		VERBOSE_FLAG="-v"
		;;
	f)
		FORCE=1
		FORCE_FLAG="-f"
		;;
	p)
		PROFILE=1
		;;
	c)
		ZPOOL_CONFIG=${OPTARG}
		;;
	t)
		ZPIOS_TEST_ARG=${OPTARG}
		ZPIOS_TEST=${ZPIOSDIR}/${OPTARG}.sh
		;;
	o)
		ZPIOS_OPTIONS=${OPTARG}
		;;
	l)	# Passed through to zpool-create.sh 
		ZPOOL_OPTIONS=${OPTARG}
		;;
	s)	# Passed through to zpool-create.sh
		ZFS_OPTIONS=${OPTARG}
		;;
	?)
		usage
		exit
		;;
	esac
done

if [ "$(id -u)" != 0 ]; then
        die "Must run as root"
fi

# Validate and source your test config
check_test || die "${ERROR}"
. "${ZPIOS_TEST}"

# Pull in the zpios test module if not loaded.  If this fails, it is
# likely because the full module stack was not yet loaded with zfs.sh
if check_modules; then
	if ! load_modules; then
		die "Run 'zfs.sh' to ensure the full module stack is loaded"
	fi
fi

# Wait for device creation
while [ ! -c /dev/zpios ]; do
	sleep 1
done

if [ ${VERBOSE} ]; then
	print_header
	print_spl_info
	print_zfs_info
fi

# Create the zpool configuration
"${ZPOOL_CREATE_SH}" "${VERBOSE_FLAG}" "${FORCE_FLAG}" \
	-p "${ZPOOL_NAME}" -c "${ZPOOL_CONFIG}" \
	-l "${ZPOOL_OPTIONS}" -s "${ZFS_OPTIONS}" || unload_die

if [ "${PROFILE}" ]; then
	zpios_profile_start
fi

zpios_start
zpios_stop

if [ ${PROFILE} ]; then
	zpios_profile_stop
fi

if [ ${VERBOSE} ]; then
	print_stats
fi

# Destroy the zpool configuration
"${ZPOOL_CREATE_SH}" "${VERBOSE_FLAG}" "${FORCE_FLAG}" \
	-p "${ZPOOL_NAME}" -c "${ZPOOL_CONFIG}" -d || unload_die

# Unload the test module stack and wait for device removal
unload_modules
while [ -c /dev/zpios ]; do
	sleep 1
done

exit 0
